### PR TITLE
Add document transformation capability for indexing

### PR DESCRIPTION
**Pull request by tharropoulos**
### TLDR
New options to transform Firestore documents before indexing to Typesense.

#### Added Configuration:
1. **In `extension.yaml`**:
   - Added 4 new configuration parameters:
     - `TRANSFORM_FUNCTION_NAME`: Name of Cloud Function for document transformation
     - `TRANSFORM_FUNCTION_PROJECT_ID`: Project where transform function is deployed
     - `TRANSFORM_FUNCTION_REGION`: Region of transform function
     - `TRANSFORM_FUNCTION_SECRET`: Auth secret for transform function

#### Added Functionality:
1. **In `utils.js`**:
   - Added `transformDocument()`: Calls external transform function with error handling
   - Implements fallback to original document when transformation fails

2. **In `indexOnWrite.js`**:
   - Updated to conditionally use document transformation before indexing
   - Added branch to handle transformed vs. non-transformed document flow

#### Added Tests:
1. **New file `test/utilsTransform.spec.js`**:
   - Tests for transform function with various scenarios:
     - Success path with proper transformation
     - Handling missing transform function configuration
     - Error handling for failed transformations
     - Edge cases like documents without IDs

2. **In `package.json`**:
   - Added new test command: `test:utils`

## PR Checklist

- [x] I have read and signed the [Contributor License Agreement](https://forms.gle/PZyiY5N2GDQU8GsV9).

### CODE REVIEW

**Review by jasonbosco — CHANGES_REQUESTED** [review]

File: functions/src/utils.js [hunk 1]
```diff
@@ -223,3 +223,50 @@ exports.pathMatchesSelector = function (path, selector) {
...
+
+  try {
+    const projectId = config.transformFunctionProjectId;
+    const region = config.transformFunctionRegion;
+
+    debug(`Calling transform function: ${transformFunctionName}`);
+
+    const url = `https://${region}-${projectId}.cloudfunctions.net/${transformFunctionName}`;
+
+    const response = await fetch(url, {
```
[the remark this answers is not in the document]
Comment (jasonbosco) [reply functions/src/utils.js:249]: @tharropoulos Looks like this is not addressed yet

File: functions/src/config.js [hunk 2]
```diff
@@ -13,4 +13,8 @@ module.exports = {
   typesenseAPIKey: process.env.TYPESENSE_API_KEY,
   typesenseBackfillTriggerDocumentInFirestore: "typesense_sync/backfill",
   typesenseBackfillBatchSize: 1000,
+  transformFunctionName: process.env.TRANSFORM_FUNCTION_NAME,
+  transformFunctionSecret: process.env.TRANSFORM_FUNCTION_SECRET,
+  transformFunctionProjectId: process.env.GCLOUD_PROJECT,
+  transformFunctionRegion: process.env.LOCATION,
```
Comment (jasonbosco) [inline functions/src/config.js:19]: Are the above two lines still needed?

**Review by jasonbosco — CHANGES_REQUESTED** [review]

File: functions/src/config.js (hunk 2 above)
Comment (tharropoulos) [reply functions/src/config.js:19]: We have to call out to eh region-project id. Both of those are coming from env variables. The gcloud project is [automatic](https://firebase.google.com/docs/functions/config-env?gen=2nd#reserved-names)  and the location is being set from `extension.yaml`.
